### ISSUE TITLE
Typescript does not check type of mapped variable correctly.

### ISSUE THREAD

**Issue by McDic** [microsoft/TypeScript#34991]
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:** ts2531 possibly null

**Code**

```ts
interface test2{
    [key: string]: number | null
}

var obj: test2 = {
    aaa: 1234,
    bbb: null,
    ccc: -5.67
}

function check2(obj: test2, key: string){
    if(typeof obj[key] === 'number' && obj[key] !== null){
        console.log(`ok obj[key] + 10 = ${obj[key] + 10}`)
    }
    else{
        console.log('ok otherwise')
    }
}

for(const key in obj){
    check2(obj, key)
}

var a: unknown;

function check3(a: unknown): any{
    if(typeof a === 'number'){
        return a + 10;
    }
}
```

**Expected behavior:** No error

**Actual behavior:** TS2531 happened on `function check2`'s console logging code. (`${obj[key] + 10}`) 

**Playground Link:** [Here](http://www.typescriptlang.org/play/?ts=3.8.0-dev.20191105&ssl=1&ssc=1&pln=30&pc=2#code/JYOwLgpgTgZghgYwgAkgZzAJgN4ChkHIDaA1hAJ4BcyGUoA5gLrUgCuAtgEbTIA+ybADaDcAX1y4AbnCjIA9pwBW1dFmQBeZHkLI4e6gEZMAZgAsAGnyFONlq2GWdCZ9QC0AVgB0ANgDsYiRhWEAQwYDkQZAQACwgEEkwACgVlVAgMTHNkMioaMDoQegBKbUJgGESwcgAHCDkYeSVSCkYNdU0AcjYuaA7kADJ+xsVm8laAQnaBe0ESqx0CBAi0OUEIT0E5ekSAAzkSYdHWgGpkAwAGDWQAEmwUo+RTi9EdovnkcR0IQTQIUoWostVutNtsOvt5GBYlAAO7AX4dN46cTiXAwORQRJLEAYbIUZCgYZzJyxeJJFJZHJvVHSWRwajBEggOQwkAAbkCwVC4UiMTiJGMiXpyEZzNZRWocBA5H+BIqVVq9V0bU63W4UERsp0UAgYFYUEicEeZ3OHORYiAA)

**Related Issues:** https://github.com/microsoft/TypeScript/issues/28131

**Comment by MartinJohns** [comment]
Duplicate of #10530. Type narrowing does not occur for indexed access forms `e[k]` where `k` is not a literal.

You can solve it by just storing the value in a variable:

    const value = obj[key];
    if(typeof value === 'number' && value !== null){
        console.log(`ok obj[key] + 10 = ${value + 10}`)
    }

**Comment by fatcerberus** [comment]
> Type narrowing does not occur for indexed access forms `e[k]` where `k` is not a literal.

See also #31445 and #34867.  At present, the compiler can't safely narrow these expressions because it only cares about the types and doesn't track the specific `k` variable used.
